pkg/dogstatsd/server.go

```diff
@@ -341,16 +342,47 @@ func (s *Server) Flush(waitForSerializer bool) {
 	s.aggregator.Flush(time.Now().Add(time.Second*10), true)
 }
 
+// dropCR drops a terminal \r from the data.
+func dropCR(data []byte) []byte {
+	if len(data) > 0 && data[len(data)-1] == '\r' {
+		return data[0 : len(data)-1]
+	}
+	return data
+}
+
+// ScanLines is an almost identical reimplementation of bufio.ScanLines, but also
+// reports if the returned line is newline-terminated
+func ScanLines(data []byte, atEOF bool) (advance int, token []byte, eol bool, err error) {
+	if atEOF && len(data) == 0 {
+		return 0, nil, false, nil
+	}
+	if i := bytes.IndexByte(data, '\n'); i >= 0 {
+		// We have a full newline-terminated line.
+		return i + 1, dropCR(data[0:i]), true, nil
+	}
+	// If we're at EOF, we have a final, non-terminated line. Return it.
+	if atEOF {
+		return len(data), dropCR(data), false, nil
+	}
+	// Request more data.
+	return 0, nil, false, nil
+}
+
 func nextMessage(packet *[]byte) (message []byte) {
 	if len(*packet) == 0 {
 		return nil
 	}
 
-	advance, message, err := bufio.ScanLines(*packet, true)
+	advance, message, eol, err := ScanLines(*packet, true)
 	if err != nil {
 		return nil
 	}
 
+	if config.Datadog.GetBool("dogstatsd_eol_required") && !eol {
```

Comment:
Isn't this `config` call a bit costly in this part of the code?

Reply:
It is! I actually had a fix for this to avoid hitting viper constantly.